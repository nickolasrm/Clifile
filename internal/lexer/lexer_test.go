package lexer_test

import (
	"testing"

	"github.com/gkampitakis/go-snaps/snaps"
	"github.com/nickolasrm/clifile/internal/lexer"
	. "github.com/nickolasrm/clifile/test/util"
	. "github.com/onsi/ginkgo/v2"
	. "github.com/onsi/gomega"
)

func TestLexer(t *testing.T) {
	RegisterFailHandler(Fail)
	RunSpecs(t, "Lexer Suite")
}

// SnapshotTokens is a helper function to test the code.
// It receives a code and tries to match to the previously saved snapshots.
// If it does not match, it will fail the test.
// If an error occurs, it will fail the test.
func SnapshotTokens(code string) {
	t := GinkgoT()
	tokens, error := lexer.Lex(code)
	Expect(error).To(BeNil())
	snaps.MatchSnapshot(t, tokens)
}

var _ = Describe("Lexer", func() {
	Describe("Lex", func() {
		When("a comment is passed", func() {
			It("should return a comment token", func() {
				SnapshotTokens("# This is a comment")
			})
			When("# is present", func() {
				It("should return a single comment token", func() {
					SnapshotTokens("# This is a #comment")
				})
			})
			When("another comment is passed", func() {
				It("should return two comment tokens", func() {
					SnapshotTokens("# C1\n# C2")
				})
			})
		})
		When("a docstring is passed", func() {
			It("should return a docstring token", func() {
				SnapshotTokens("## This is a docstring")
			})
			When("## is present", func() {
				It("should return a single docstring token", func() {
					SnapshotTokens("## This is a ##docstring")
				})
			})
			When("another docstring is passed", func() {
				It("should return two docstring tokens", func() {
					SnapshotTokens("## D1\n## D2")
				})
			})
		})
		When("a variable is passed", func() {
			When("unquoted value is passed", func() {
				It("should return a variable token", func() {
					SnapshotTokens("VAR=val")
				})
				When("multiple lines are passed", func() {
					It("should return only the variable line", func() {
						SnapshotTokens("VAR=val\n#comment")
					})
				})
			})
			When("quoted value is passed", func() {
				It("should return a variable token", func() {
					SnapshotTokens("VAR=\"val\"")
				})
				When("multiple lines are between quotes", func() {
					It("should return a variable token", func() {
						SnapshotTokens("VAR=\"val\nval\"")
					})
				})
				When("space or tabs appear around equal", func() {
					It("should return a variable token", func() {
						SnapshotTokens("VAR \t=  \t\"ASD\"")
					})
				})
			})
		})
		When("a call is passed", func() {
			It("should return a call token", func() {
				SnapshotTokens("VAR=${func}")
			})
			When("multiple lines are between curly braces", func() {
				It("should return a call token", func() {
					SnapshotTokens("VAR=${func\nfunc}")
				})
			})
			When("space or tabs appear around equal", func() {
				It("should return a call token", func() {
					SnapshotTokens("VAR\t =  \t${func}")
				})
			})
		})
		When("a rule is passed", func() {
			It("should return a rule token", func() {
				SnapshotTokens("rule:")
			})
			When("arguments are passed", func() {
				It("should return a rule token", func() {
					SnapshotTokens("rule: arg1 arg2")
				})
			})
			When("another rule is passed", func() {
				It("should return two rule tokens", func() {
					SnapshotTokens("rule1: a b\nrule2: c d")
				})
				When("the second rule is nested", func() {
					It("should return two rule tokens", func() {
						SnapshotTokens("rule1: a\n\trule2:")
					})
				})
			})
		})
		When("an action is passed", func() {
			It("should return an action token", func() {
				SnapshotTokens("echo \"Hello World\"")
			})
			When("multiple lines are passed", func() {
				It("should return multiple action tokens", func() {
					SnapshotTokens("echo \"Hello World\"\necho \"Hello World\"")
				})
			})
		})
		When("a code snippet is passed", func() {
			It("should return a sequence of tokens", func() {
				SnapshotTokens(`
# This is a comment
VAR=val
VAR2="val"
VAR3=${func}
VAR4=${func
func}
rule1: a b
	rule4:
		echo "Hello World"
rule2: c d
## This is a docstring
rule3:
	echo "Hello World"
	echo "Hello World"
`)
			})
		})
	})

	Describe("Match", func() {
		When("a new match is created", func() {
			var m *lexer.Match
			BeforeEach(func() {
				m = lexer.NewMatch(lexer.Action, []string{"echo Hello World"})
			})

			When("Type() is called", func() {
				It("should return the type of the match", func() {
					MatchSnapshot(m.Type())
				})
			})

			When("Value() is called", func() {
				When("the index exists", func() {
					It("should return the first value of the match", func() {
						MatchSnapshot(m.Value(0))
					})
				})
				When("the index is out of bounds", func() {
					It("should return an empty string", func() {
						Expect(func() { m.Value(1) }).To(Panic())
					})
				})
			})
		})
	})
})
