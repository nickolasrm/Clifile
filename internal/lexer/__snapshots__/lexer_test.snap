
[Lexer when a comment is passed should return a comment token - 1]
[]*lexer.Match{
    &lexer.Match{
        Type:  0x3,
        Value: {"# This is a comment", " This is a comment"},
    },
}
---

[Lexer when a comment is passed when # is present should return a single comment token - 1]
[]*lexer.Match{
    &lexer.Match{
        Type:  0x3,
        Value: {"# This is a #comment", " This is a #comment"},
    },
}
---

[Lexer when a comment is passed when another comment is passed should return two comment tokens - 1]
[]*lexer.Match{
    &lexer.Match{
        Type:  0x3,
        Value: {"# C1", " C1"},
    },
    &lexer.Match{
        Type:  0x0,
        Value: {"\n"},
    },
    &lexer.Match{
        Type:  0x3,
        Value: {"# C2", " C2"},
    },
}
---

[Lexer when a docstring is passed should return a docstring token - 1]
[]*lexer.Match{
    &lexer.Match{
        Type:  0x2,
        Value: {"## This is a docstring", " This is a docstring"},
    },
}
---

[Lexer when a docstring is passed when ## is present should return a single docstring token - 1]
[]*lexer.Match{
    &lexer.Match{
        Type:  0x2,
        Value: {"## This is a ##docstring", " This is a ##docstring"},
    },
}
---

[Lexer when a docstring is passed when another docstring is passed should return two docstring tokens - 1]
[]*lexer.Match{
    &lexer.Match{
        Type:  0x2,
        Value: {"## D1", " D1"},
    },
    &lexer.Match{
        Type:  0x0,
        Value: {"\n"},
    },
    &lexer.Match{
        Type:  0x2,
        Value: {"## D2", " D2"},
    },
}
---

[Lexer when a variable is passed when unquoted value is passed should return a variable token - 1]
[]*lexer.Match{
    &lexer.Match{
        Type:  0x5,
        Value: {"VAR=val", "VAR", "", "val"},
    },
}
---

[Lexer when a variable is passed when unquoted value is passed when multiple lines are passed should return only the variable line - 1]
[]*lexer.Match{
    &lexer.Match{
        Type:  0x5,
        Value: {"VAR=val", "VAR", "", "val"},
    },
    &lexer.Match{
        Type:  0x0,
        Value: {"\n"},
    },
    &lexer.Match{
        Type:  0x3,
        Value: {"#comment", "comment"},
    },
}
---

[Lexer when a variable is passed when quoted value is passed should return a variable token - 1]
[]*lexer.Match{
    &lexer.Match{
        Type:  0x5,
        Value: {"VAR=\"val\"", "VAR", "val", ""},
    },
}
---

[Lexer when a variable is passed when quoted value is passed when multiple lines are between quotes should return a variable token - 1]
[]*lexer.Match{
    &lexer.Match{
        Type:  0x5,
        Value: {"VAR=\"val\nval\"", "VAR", "val\nval", ""},
    },
}
---

[Lexer when a call is passed should return a call token - 1]
[]*lexer.Match{
    &lexer.Match{
        Type:  0x4,
        Value: {"VAR=${func}", "VAR", "func", ""},
    },
}
---

[Lexer when a call is passed when multiple lines are between curly braces should return a call token - 1]
[]*lexer.Match{
    &lexer.Match{
        Type:  0x4,
        Value: {"VAR=${func\nfunc}", "VAR", "func", "func"},
    },
}
---

[Lexer when a rule is passed should return a rule token - 1]
[]*lexer.Match{
    &lexer.Match{
        Type:  0x6,
        Value: {"rule:", "rule", ""},
    },
}
---

[Lexer when a rule is passed when arguments are passed should return a rule token - 1]
[]*lexer.Match{
    &lexer.Match{
        Type:  0x6,
        Value: {"rule: arg1 arg2", "rule", " arg1 arg2"},
    },
}
---

[Lexer when a rule is passed when another rule is passed should return two rule tokens - 1]
[]*lexer.Match{
    &lexer.Match{
        Type:  0x6,
        Value: {"rule1: a b", "rule1", " a b"},
    },
    &lexer.Match{
        Type:  0x0,
        Value: {"\n"},
    },
    &lexer.Match{
        Type:  0x6,
        Value: {"rule2: c d", "rule2", " c d"},
    },
}
---

[Lexer when a rule is passed when another rule is passed when the second rule is nested should return two rule tokens - 1]
[]*lexer.Match{
    &lexer.Match{
        Type:  0x6,
        Value: {"rule1: a", "rule1", " a"},
    },
    &lexer.Match{
        Type:  0x0,
        Value: {"\n"},
    },
    &lexer.Match{
        Type:  0x1,
        Value: {"\t"},
    },
    &lexer.Match{
        Type:  0x6,
        Value: {"rule2:", "rule2", ""},
    },
}
---

[Lexer when an action is passed should return an action token - 1]
[]*lexer.Match{
    &lexer.Match{
        Type:  0x7,
        Value: {"echo \"Hello World\"", "echo \"Hello World\""},
    },
}
---

[Lexer when an action is passed when multiple lines are passed should return multiple action tokens - 1]
[]*lexer.Match{
    &lexer.Match{
        Type:  0x7,
        Value: {"echo \"Hello World\"", "echo \"Hello World\""},
    },
    &lexer.Match{
        Type:  0x0,
        Value: {"\n"},
    },
    &lexer.Match{
        Type:  0x7,
        Value: {"echo \"Hello World\"", "echo \"Hello World\""},
    },
}
---

[Lexer when a code snippet is passed should return a sequence of tokens - 1]
[]*lexer.Match{
    &lexer.Match{
        Type:  0x0,
        Value: {"\n"},
    },
    &lexer.Match{
        Type:  0x3,
        Value: {"# This is a comment", " This is a comment"},
    },
    &lexer.Match{
        Type:  0x0,
        Value: {"\n"},
    },
    &lexer.Match{
        Type:  0x5,
        Value: {"VAR=val", "VAR", "", "val"},
    },
    &lexer.Match{
        Type:  0x0,
        Value: {"\n"},
    },
    &lexer.Match{
        Type:  0x5,
        Value: {"VAR2=\"val\"", "VAR2", "val", ""},
    },
    &lexer.Match{
        Type:  0x0,
        Value: {"\n"},
    },
    &lexer.Match{
        Type:  0x4,
        Value: {"VAR3=${func}", "VAR3", "func", ""},
    },
    &lexer.Match{
        Type:  0x0,
        Value: {"\n"},
    },
    &lexer.Match{
        Type:  0x4,
        Value: {"VAR4=${func\nfunc}", "VAR4", "func", "func"},
    },
    &lexer.Match{
        Type:  0x0,
        Value: {"\n"},
    },
    &lexer.Match{
        Type:  0x6,
        Value: {"rule1: a b", "rule1", " a b"},
    },
    &lexer.Match{
        Type:  0x0,
        Value: {"\n"},
    },
    &lexer.Match{
        Type:  0x1,
        Value: {"\t"},
    },
    &lexer.Match{
        Type:  0x6,
        Value: {"rule4:", "rule4", ""},
    },
    &lexer.Match{
        Type:  0x0,
        Value: {"\n"},
    },
    &lexer.Match{
        Type:  0x1,
        Value: {"\t\t"},
    },
    &lexer.Match{
        Type:  0x7,
        Value: {"echo \"Hello World\"", "echo \"Hello World\""},
    },
    &lexer.Match{
        Type:  0x0,
        Value: {"\n"},
    },
    &lexer.Match{
        Type:  0x6,
        Value: {"rule2: c d", "rule2", " c d"},
    },
    &lexer.Match{
        Type:  0x0,
        Value: {"\n"},
    },
    &lexer.Match{
        Type:  0x2,
        Value: {"## This is a docstring", " This is a docstring"},
    },
    &lexer.Match{
        Type:  0x0,
        Value: {"\n"},
    },
    &lexer.Match{
        Type:  0x6,
        Value: {"rule3:", "rule3", ""},
    },
    &lexer.Match{
        Type:  0x0,
        Value: {"\n"},
    },
    &lexer.Match{
        Type:  0x1,
        Value: {"\t"},
    },
    &lexer.Match{
        Type:  0x7,
        Value: {"echo \"Hello World\"", "echo \"Hello World\""},
    },
    &lexer.Match{
        Type:  0x0,
        Value: {"\n"},
    },
    &lexer.Match{
        Type:  0x1,
        Value: {"\t"},
    },
    &lexer.Match{
        Type:  0x7,
        Value: {"echo \"Hello World\"", "echo \"Hello World\""},
    },
    &lexer.Match{
        Type:  0x0,
        Value: {"\n"},
    },
}
---
